src/probnum/diffeq/odesolution.py

```diff
@@ -1,176 +1,70 @@
-        return self._kalman_posterior.state_rvs
-
-    def __call__(self, t):
+    def dy(self) -> pnrv_list._RandomVariableList:
+        """First derivative of the discrete-time solution."""
+        raise NotImplementedError("The first derivative has not been implemented")
+
+    # Not abstract, because providing interpolation could sometimes be tedious.
+    def __call__(
+        self, t: float
+    ) -> typing.Union[pnrv.RandomVariable, pnrv_list._RandomVariableList]:
         """Evaluate the time-continuous solution at time t.
 
-        `KalmanPosterior.__call__` does the main algorithmic work to return the
-        posterior for a given location. All that is left to do here is to (1) undo the
-        preconditioning, and (2) to slice the state_rv in order to return only the
-        rv for the function value.
-
         Parameters
         ----------
-        t : float
+        t
             Location / time at which to evaluate the continuous ODE solution.
 
         Returns
         -------
-        :obj:`RandomVariable`
-            Probabilistic estimate of the continuous-time solution at time ``t``.
+        Probabilistic estimate of the continuous-time solution at time ``t``.
         """
-        out_rv = self._kalman_posterior(t)
-        projmat = self._solver.prior.proj2coord(coord=0)
+        raise NotImplementedError("Dense output is not implemented.")
 
-        if np.isscalar(t):
-            return projmat @ out_rv
-
-        return _RandomVariableList([projmat @ rv for rv in out_rv])
-
-    def __len__(self):
+    def __len__(self) -> int:
         """Number of points in the discrete-time solution."""
-        return len(self._kalman_posterior)
-
-    def __getitem__(self, idx):
-        """Access the discrete-time solution through indexing and slicing."""
-        projmat = self._solver.prior.proj2coord(coord=0)
-
-        if isinstance(idx, int):
-            rv = self._kalman_posterior[idx]
-            return projmat @ rv
-        elif isinstance(idx, slice):
-            rvs = self._kalman_posterior[idx]
-            f_rvs = [projmat @ rv for rv in rvs]
-            return _RandomVariableList(f_rvs)
-        else:
-            raise ValueError("Invalid index")
-
-    def sample(self, t=None, size=()):
-
-        size = utils.as_shape(size)
-        projmat = self._solver.prior.proj2coord(coord=0)
-
-        # implement only single samples, rest via recursion
-        if size != ():
-            return np.array([self.sample(t=t, size=size[1:]) for _ in range(size[0])])
-
-        samples = self._kalman_posterior.sample(locations=t, size=size)
-        return np.array([projmat @ sample for sample in samples])
+        return len(self.y)
+
+    def __getitem__(self, idx: int) -> pnrv.RandomVariable:
+        """Access the :math:`i`th element of the discrete-time solution."""
+        return self.y[idx]
+
+    def sample(
+        self,
+        t: typing.Optional[float] = None,
```

Comment:
Here I'm wondering again what the function _should_ actually do, and in particular about the role of `t`. What should be sampled here?

If you have a clear idea, it would be great if you could formulate that as a docstring and add it here, otherwise we can discuss what behaviour we want here.

Reply:
I forgot the Union again. Will also explain in the docstring what happens with different inputs.